cip/1.accepted/CIP2015-05-13-EXISTS.adoc

```diff
@@ -4,102 +4,185 @@
+https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>], <SimpleExistentialSubquery>, and <FullExistentialSubquery> accomplish the same task: checking whether the result of a subquery is not empty.
+
+All three forms,
+
+* Are boolean expressions, i.e. return a single boolean value
+* Never return `null`
+* Are side effect-free, i.e. <FullExistentialSubquery> shall not contain any <UpdatingClause> or other sources of side effects
+* Can contain variables and parameters from the outer queries
 
-==== Property exists
+https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] shall not to introduce new variables.
 
-In the property checking form, any expression is accepted as the input.
-If the input expression evaluates to `null`, `exists()` will evaluate to `false`. For any other value, it will evaluate to `true`.
-This form of `exists()` is equivalent to checking for `null` using `IS NOT NULL`: `exists(expr)` is equivalent to `expr IS NOT NULL`.
+Both forms of https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=ExistentialSubquery[<ExistentialSubquery>] (<SimpleExistentialSubquery> and <FullExistentialSubquery>) are allowed to introduce new variables.
+These variables necessarily shall have a name different from the names of all variables available from the outer queries.
+Any variables introduced in an <ExistentialSubquery> are not available on the outside.
 
-This form is there to make property existence checking idiomatic in Cypher - we do not want the user thinking of missing properties as checking for `null` values.
+https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] and <SimpleExistentialSubquery> are syntactically simpler and semantically less powerful forms of <FullExistentialSubquery>.
+The semantics of https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] and <SimpleExistentialSubquery> can be defined as syntax transformations to <FullExistentialSubquery>.
 
-==== Subquery exists
+==== <RelationshipsPattern>
 
-When checking for the existence of subqueries, variables and parameters from the outer query are available in the subquery.
-Variables introduced in the subquery are not available on the outside.
+A https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] _RP_ is effectively replaced by the expression
 
-For each matching subgraph evaluated with `EXISTS {}`, the result value must be `true` if the subquery finds at least one matching row.
-If no matches are found, `false` should be returned.
+`EXISTS { MATCH _RP_ RETURN 1 }`
```

Comment:
RETURN 1 is arbitrary, but I don't know of a less arbitrary alternative. I considered commenting we should write `RETURN true` here instead to stay with booleans, but perhaps that could be misinterpreted as being an important return value. No return value is important; the presence of any row indicates existence. It would be clean to have some kind of "return a row with no fields" construct.

`RETURN *` also does not work since RPs cannot introduce variables so it could fail.

Reply:
Good observation and thanks for running through a number of alternative. I hadn't really do that.

Also: It just specifies what is effectively suppose to happen. How an implementation precisely does it is up to the implementation. Hence, it is also not super important which of the arbitrary values the specification uses as long as it is not wrong or terribly misleading.

Reply:
Agreed. But the ideal thing would be for the spec to mention nothing that is irrelevant; unfortunately we are unable to achieve that right here -- your choice is the best though.